C error when trying to get sum type value from map

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Linux version 6.0.8-arch1-1 (linux@archlinux) (gcc (GCC) 12.2.0, GNU ld (GNU Binutils) 2.39.0) #1 SMP PREEMPT_DYNAMIC Thu, 10 Nov 2022 21:14:24 +0000
Processor: 4 cpus, 64bit, little endian, Intel(R) Pentium(R) CPU 4415Y @ 1.60GHz
CC version: cc (GCC) 12.2.0

getwd: /home/archie/projects/v-playground/map-sumtype
vmodules: /home/archie/.vmodules
vroot: /home/archie/projects/v
vexe: /home/archie/projects/v/v
vexe mtime: 2022-11-14 12:22:14
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.2 12f0131.1a4d1ae

Git version: git version 2.38.1
Git vroot status: weekly.2022.13-1730-g1a4d1aec
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg mapsumtype.v`
```v
module main

type ConfigValue = bool | int | string
type ConfigMap = map[string]ConfigValue

fn foo(conf ConfigMap)
{
	mut bar := false
	bar =
		if conf['baz'] or { false } is bool // Check type
		{ conf['baz'] or { false } as bool }
		else
		{ false } // Default value
}

fn main()
{
	conf := { 'baz': ConfigValue(123) }
	foo(conf)
}

```

**What did you expect to see?**

Successful compilation


**What did you see instead?**
```
mapsumtype.v:8:6: warning: unused variable: `bar`
    6 | fn foo(conf ConfigMap)
    7 | {
    8 |     mut bar := false
      |         ~~~
    9 |     bar =
   10 |         if conf['baz'] or { false } is bool // Check type
==================
/tmp/v_1000/mapsumtype.18309332128107798760.tmp.c:12163: error: ')' expected (got "_t4")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -11,7 +11,10 @@ fn (mut g Gen) need_tmp_var_in_if(node ast.IfExpr) bool {
 			return true
 		}
 		for branch in node.branches {
-			if branch.cond is ast.IfGuardExpr || branch.stmts.len > 1 {
+			if branch.stmts.len > 1 {
+				return true
+			}
+			if g.need_tmp_var_in_expr(branch.cond) {
 				return true
 			}
 			if branch.stmts.len == 1 {
@@ -34,6 +37,17 @@ fn (mut g Gen) need_tmp_var_in_expr(expr ast.Expr) bool {
 	match expr {
 		ast.IfExpr {
 			if g.need_tmp_var_in_if(expr) {
+				return true
+			}
+		}
+		ast.IfGuardExpr {
+			return true
+		}
+		ast.InfixExpr {
+			if g.need_tmp_var_in_expr(expr.left) {
+				return true
+			}
+			if g.need_tmp_var_in_expr(expr.right) {
 				return true
 			}
 		}

--- a/vlib/v/tests/if_expr_with_sumtype_map_test.v
+++ b/vlib/v/tests/if_expr_with_sumtype_map_test.v
@@ -1,0 +1,24 @@
+module main
+
+type ConfigValue = bool | int | string
+type ConfigMap = map[string]ConfigValue
+
+fn foo(conf ConfigMap) bool {
+	mut bar := false
+	// Check type
+	bar = if conf['baz'] or { false } is bool {
+		conf['baz'] or { false } as bool
+	} else {
+		false
+	} // Default value
+	return bar
+}
+
+fn test_if_expr_with_sumtype_map() {
+	conf := {
+		'baz': ConfigValue(123)
+	}
+	ret := foo(conf)
+	println(ret)
+	assert !ret
+}

